src/fdcache.cpp

```diff
@@ -2181,17 +2195,22 @@ bool FdManager::ChangeEntityToTempPath(FdEntity* ent, const char* path)
 
 void FdManager::CleanupCacheDir()
 {
-  if (!FdManager::IsCacheDir()) {
+  S3FS_PRN_INFO("cache cleanup requested");
+
+  if(!FdManager::IsCacheDir()){
     return;
   }
 
-  AutoLock auto_lock(&FdManager::cache_cleanup_lock, true);
+  AutoLock auto_lock_no_wait(&FdManager::cache_cleanup_lock, true);
 
-  if (!auto_lock.isLockAcquired()) {
-    return;
+  if(auto_lock_no_wait.isLockAcquired()){
+    S3FS_PRN_INFO("cache cleanup started");
+    CleanupCacheDirInternal("");
+    S3FS_PRN_INFO("cache cleanup ended");
+  }else{
+    // wait for other thread to finish cache cleanup
+    AutoLock auto_lock(&FdManager::cache_cleanup_lock);
   }
-
-  CleanupCacheDirInternal("");
 }
 
 void FdManager::CleanupCacheDirInternal(const std::string &path)
```

Comment:
Hey @orozery when cleaning up the cache, where do we update the free_disk_space?  I looked through the code and its not clear to me.  I see that we initialize the cache with size 0, but I can't find the place where we reduce it during a cleanup while keeping the amount of reserved space accounted for.

Reply:
free_disk_space is actually reserved_disk_space.
To check if you have enough space, you check GetDiskFreeSpace() - reserved_disk_space > requested_amount